eZ/Publish/Core/Helper/PreviewLocationProvider.php

```diff
@@ -0,0 +1,87 @@
+<?php
+/**
+ * This file is part of the eZ Publish Kernel package
+ *
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributd with this source code.
+ */
+namespace eZ\Publish\Core\Helper;
+
+use eZ\Publish\API\Repository\ContentService;
+use eZ\Publish\API\Repository\LocationService;
+use eZ\Publish\Core\Repository\Values\Content\Location;
+use eZ\Publish\SPI\Persistence\Handler as PersistenceHandler;
+
+/**
+ * Provides location(s) for a content. Handles unpublished content that does not have an actual location yet.
+ */
+class PreviewLocationProvider
+{
+    /** @var \eZ\Publish\API\Repository\LocationService */
+    private $locationService;
+
+    /** @var \eZ\Publish\API\Repository\ContentService */
+    private $contentService;
+
+    /** @var \eZ\Publish\SPI\Persistence\Content\Location\Handler */
+    private $locationHandler;
+
+    /**
+     * @param \eZ\Publish\API\Repository\LocationService $locationService
+     * @param \eZ\Publish\API\Repository\ContentService $contentService
+     * @param \eZ\Publish\SPI\Persistence\Handler $persistenceHandler
+     */
+    public function __construct(
+        LocationService $locationService,
+        ContentService $contentService,
+        PersistenceHandler $persistenceHandler
+    )
+    {
+        $this->locationService = $locationService;
+        $this->contentService = $contentService;
+        $this->locationHandler = $persistenceHandler->locationHandler();
+    }
+
+    /**
+     * Loads the main location for $contentId
+     *
+     * If the content does not have a location (yet), an Location draft is returned.
+     * Location drafts do not have an id (it is set to null), and can be tested using the isDraft() method.
+     *
+     * @param mixed $contentInfo
+     *
+     * @return \eZ\Publish\API\Repository\Values\Content\Location
+     */
+    public function loadMainLocation( $contentId )
+    {
+        $contentInfo = $this->contentService->loadContentInfo( $contentId );
+
+        // mainLocationId already exists, content has been published at least once.
+        if ( $contentInfo->mainLocationId )
```

Comment:
Might not have been :p

Note that I like the three stages Brandon Savage described in his article about reviews: "you might", "you should", "you must". In any case, we can always handle this one in another commit, before backporting if necessary.